### PR TITLE
Slack emoji support

### PR DESCRIPTION
**Pull request by MeijerM1**
- Users can like/dislike kudo-o-matic message posts using emojis
- Users can create posts using emojis
- Connecting Kudo-O-Matic accounts to Slack is now done via OAuth
- Slack events are handled asynchronous with Sidekiq to comply with the [Slack events API](https://api.slack.com/events-api#responding_to_events)

A good next step would be to refactor the SlackService.rb class into smaller modules like auth, message_parsing, etc. 

example gif

### CODE REVIEW

File: app/controllers/slack_controller.rb [hunk 1]
```diff
@@ -40,22 +40,23 @@ def guidelines
     end
   end
 
-  def auth
-    redirect_to SlackService.get_oauth_url(params[:team_id]).to_s
+  def auth_team
+    redirect_to SlackService.get_team_oauth_url(params[:team_id]).to_s
```
Comment (thnukid) [inline app/controllers/slack_controller.rb]: You can remove '.to_s' or with other words `get_team_oauth_url` should return a string, so that you do not need to call '.to_s' on that.

File: app/controllers/slack_controller.rb [hunk 2]
```diff
@@ -40,22 +40,23 @@ def guidelines
...
   end
 
-  def reaction
+  def auth_user
+    redirect_to SlackService.get_user_oauth_url(params[:user_id]).to_s
```
Comment (thnukid) [inline app/controllers/slack_controller.rb]: same,  remove `to_s`

File: app/controllers/slack_controller.rb [hunk 3]
```diff
@@ -40,22 +40,23 @@ def guidelines
...
     case params[:type]
     when 'url_verification'
-      render json: {challenge: params[:challenge]}
-    when 'reaction_added'
-      render json: {text: 'somebody liked something!'}
-    when 'reaction_removed'
-      render json: {text: 'somebody unliked something!'}
+      return render json: {challenge: params[:challenge]}
+    when 'event_callback'
+      SlackWorker.perform_async(params[:team_id], params[:event].to_unsafe_h)
```
Comment (thnukid) [inline app/controllers/slack_controller.rb:56]: `to_unsafe_h` is that because there are no strong parameters in the controller.

Check https://api.rubyonrails.org/classes/ActionController/Parameters.html - Strong Parameters

**Review by thnukid — COMMENTED** [review]
Good work Meijer!

File: app/controllers/slack_controller.rb [hunk 4]
```diff
@@ -13,19 +13,19 @@ def give_kudos
     end
   end
 
-  def register
+  def team_auth_callback
```
Comment (ariejan) [inline app/controllers/slack_controller.rb:16]: This does not seem related to "Slack emoji support", try to keep you PR's small and focussed, that makes reviewing and merging easier.

**Review by ariejan — APPROVED** [review]
The code looks good to me. Didn't try and run it. 

But, for the life of me, I can't find any reference to handling emoji in Slack.

File: app/controllers/slack_controller.rb (hunk 4 above)
Comment (MeijerM1) [reply app/controllers/slack_controller.rb:16]: It relates because the flow for connecting users has changed in order to facilitate the editing of messages.

Before the emoji (read 'event') support a registration token was used to connect Kudo-O-Matic accounts to Slack accounts but because the editing of Slack messages requires a Slack access token from the author of the message, OAUth is now used. In the gif you can see the 'This message is a Kudo-O-Matic' text being appended to the original Slack message. 

Therefore the auth flow needed to be redesigned.
